### PR TITLE
Implement Forget Password Flow

### LINKED ISSUE #77: E-01-S05 Password recovery.
## Feature Request

As an organization user, want to be able to request password recovery.  The user steps can be as follows.

1. The user requests password recovery via the mobile app.
2. The user will be asked to choose an option between mobile or email notifications.
3. Temporary link is sent to users with a warning of expiration.
4. If the user click on the link before expiration, the user is sent to a link to reset the password to change the password according to the password policy.
5. If the link is expired, the User will get an error message with a link to request another password recovery request.
6. Once the password is changed, the user should be redirected to the login UI

## Acceptance Criteria
- [ ] Successful password recovery for web app
- [ ] Temporary link expiration after 48 hours
- [ ] Limiting password change requests for a user up to 5 times a day. (optional)
- [ ] Warning for users if the password change request is made 4 times on that day. (optional)

## Implementation Details (optional)
AWS cognito should have native support for all these feature. Please check cognito documentation for more details.

### PR DESCRIPTION
**Pull request by EbrahimAli-RU**
Fixes #77 

Resolves #
Used when the changes in the pull request indirectly resolve another issue.

- [] Does not depend on other PRs
- [x] Depends on other PR being merged first: #197 

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [x] New feature
- [ ] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?

In the Forgot Password process, the user first requests a password reset by entering their registered Email. An OTP is then sent to the user for verification. Once the OTP is received, the user enters it along with a new password. This information is sent to the backend for validation and password update. If successful, the user is redirected to the login page to access their account with the new password.

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [x] I have followed the project's coding style and guidelines.
- [x] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [x] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [x] No risk
- [ ] Low
- [ ] High

### CODE REVIEW

**Comment by melon-ruet** [comment]
> Depends on other PR being merged first: https://github.com/CraftsmenLtd/BloodConnect/issues/185

This is not a PR link

**Comment by melon-ruet** [comment]
I can see all codes are under "mobile" directory. But forget password feature is also necessary in web app. Are we thinking about the core business logic in a single place?

**Comment by EbrahimAli-RU** [comment]
> I can see all codes are under "mobile" directory. But forget password feature is also necessary in web app. Are we thinking about the core business logic in a single place?

Yes, vai, we are thinking of sharing common functionality between mobile and web applications since the web app is paused. We have a plan to move common functionality to a common directory after the first iteration.

**Comment by EbrahimAli-RU** [comment]
> > Depends on other PR being merged first: #185
> 
> This is not a PR link

Updated.

File: clients/mobile/__tests__/__mocks__/reactNavigation.mock.ts [hunk 1]
```diff
@@ -1,19 +1,30 @@
 const mockedNavigate = jest.fn()
 const mockDispatch = jest.fn()
+const mockReset = jest.fn()
 
 jest.mock('@react-navigation/native', () => {
+  let routeParams = {}
```
Comment (sakibstark11) [inline clients/mobile/__tests__/__mocks__/reactNavigation.mock.ts:6]: I don't see why we need to do it like this, why can't we use returns where we need it?

File: clients/mobile/__tests__/authentication/authService.test.ts [hunk 2]
```diff
@@ -234,6 +238,90 @@ describe('AuthService', () => {
...
+
+      await expect(resetPasswordHandler(mockEmail)).rejects.toThrow(errorMessage)
+      expect(resetPassword).toHaveBeenCalledWith({ username: mockEmail })
+    })
+  })
+
+  describe('handleConfirmPasswordError', () => {
+    const handleConfirmPasswordError = (error: unknown): string => {
+      if (error instanceof Error) {
+        switch (true) {
```
Comment (sakibstark11) [inline clients/mobile/__tests__/authentication/authService.test.ts:263]: no reason for switch if you're using a boolean

File: clients/mobile/__tests__/authentication/authService.test.ts [hunk 3]
```diff
@@ -234,6 +238,90 @@ describe('AuthService', () => {
...
+
+    it('should return generic error message for non-Error objects', () => {
+      const error = 'Unexpected error'
+      expect(handleConfirmPasswordError(error)).toBe('An unexpected error occurred. Please try again.')
+    })
+  })
+
+  describe('confirmResetPasswordHandler', () => {
+    it('should return true on successful password confirmation', async() => {
+      ;(confirmResetPassword as jest.Mock).mockResolvedValueOnce(true)
```
Comment (sakibstark11) [inline clients/mobile/__tests__/authentication/authService.test.ts:299]: ?

File: clients/mobile/__tests__/authentication/authService.test.ts [hunk 4]
```diff
@@ -234,6 +238,90 @@ describe('AuthService', () => {
...
+      expect(confirmResetPassword).toHaveBeenCalledWith({
+        username: mockEmail,
+        confirmationCode: mockOtp,
+        newPassword: mockPassword
+      })
+    })
+
+    it('should throw a custom error message on failure', async() => {
+      const errorMessage = 'Invalid verification code'
+      ;(confirmResetPassword as jest.Mock).mockRejectedValueOnce(new Error(errorMessage))
```
Comment (sakibstark11) [inline clients/mobile/__tests__/authentication/authService.test.ts:312]: ?

File: clients/mobile/src/authentication/otp/hooks/useOtp.ts [hunk 5]
```diff
@@ -34,20 +34,28 @@ export const useOtp = (): any => {
     return otp.join('').trim().length !== 6
   }, [otp])
 
+  const handleRegister = async(): Promise<void> => {
+    const isSucessRegister = await submitOtp(email, otp.join(''))
+    if (isSucessRegister) {
```
Comment (sakibstark11) [inline clients/mobile/src/authentication/otp/hooks/useOtp.ts:39]: could fail first be a better approach?

File: clients/mobile/src/authentication/services/authService.ts [hunk 6]
```diff
@@ -139,6 +139,41 @@ export const facebookLogin = async(): Promise<void> => {
...
+    return nextStep
+  } catch (error) {
+    const errorMessage = handleAuthError(error)
+    throw new Error(errorMessage)
+  }
+}
+
+const handleConfirmPasswordError = (error: unknown): string => {
+  if (error instanceof Error) {
+    switch (true) {
```
Comment (sakibstark11) [inline clients/mobile/src/authentication/services/authService.ts:154]: same

File: clients/mobile/__tests__/__mocks__/reactNavigation.mock.ts (hunk 1 above)
Comment (EbrahimAli-RU) [reply clients/mobile/__tests__/__mocks__/reactNavigation.mock.ts]: Updated to const [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/202/commits/ba27593e9bda7d902822fc85881238bf5b32500a)

File: clients/mobile/src/authentication/services/authService.ts (hunk 6 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/authentication/services/authService.ts]: Updated  [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/202/commits/ba27593e9bda7d902822fc85881238bf5b32500a)

File: clients/mobile/src/authentication/otp/hooks/useOtp.ts (hunk 5 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/authentication/otp/hooks/useOtp.ts]: Updated  [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/202/commits/ba27593e9bda7d902822fc85881238bf5b32500a)

File: clients/mobile/__tests__/authentication/authService.test.ts (hunk 4 above)
Comment (EbrahimAli-RU) [reply clients/mobile/__tests__/authentication/authService.test.ts]: Updated  [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/202/commits/ba27593e9bda7d902822fc85881238bf5b32500a)

File: clients/mobile/__tests__/authentication/authService.test.ts (hunk 3 above)
Comment (EbrahimAli-RU) [reply clients/mobile/__tests__/authentication/authService.test.ts]: Updated  [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/202/commits/ba27593e9bda7d902822fc85881238bf5b32500a)

File: clients/mobile/__tests__/authentication/authService.test.ts (hunk 2 above)
Comment (EbrahimAli-RU) [reply clients/mobile/__tests__/authentication/authService.test.ts]: Removed switch  [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/202/commits/ba27593e9bda7d902822fc85881238bf5b32500a)

**Review by sakibstark11 — APPROVED** [review]

**Review by mahinuzzaman — APPROVED** [review]

**Pull request merged**